test/tests/alt_text_highlighting_test.py

```diff
@@ -0,0 +1,348 @@
+    Morpheme(
+        lemma="そうですね",
+        inflection="そうですね",
+        highest_inflection_learning_interval=0,
+    ),
+]
+
+##############################################################################################
+#                                CASE: Several cases that uncovered bugs in new impl
+##############################################################################################
+
+case_japanese_three_params = FakeEnvironmentParams(
+    config=config_big_japanese_collection,
+)
+CASE_JAPANESE_THREE_INPUT_TEXT = "雪[ゆき]が お 留守番[るすばん]  相変[あいか]わらずの"
+CASE_JAPANESE_THREE_CORRECT_OUTPUT = (
+    '<span morph-status="known"><ruby>雪<rt>ゆき</rt></ruby></span><span morph-status="learning">が</span>'
+    + " "
+    + '<span morph-status="learning">お</span><ruby><span morph-status="known">留守</span><span morph-status="learning">番</span><rt>るすばん</rt></ruby>'
+    + " "
+    + '<ruby><span morph-status="learning">相</span><span morph-status="learning">変</span><rt>あいか</rt></ruby><span morph-status="learning">わら</span><span morph-status="learning">ず</span>の'
+)
+
+case_japanese_three_card_morphs = [
+    Morpheme(lemma="雪", inflection="雪", highest_inflection_learning_interval=30),
+    Morpheme(lemma="が", inflection="が", highest_inflection_learning_interval=10),
+    Morpheme(lemma="留守", inflection="留守", highest_inflection_learning_interval=30),
+    Morpheme(lemma="番", inflection="番", highest_inflection_learning_interval=10),
+    Morpheme(lemma="お", inflection="お", highest_inflection_learning_interval=10),
+    Morpheme(lemma="見事", inflection="見事", highest_inflection_learning_interval=30),
+    Morpheme(lemma="腕", inflection="腕", highest_inflection_learning_interval=10),
+    Morpheme(
+        lemma="変わら", inflection="変わら", highest_inflection_learning_interval=10
+    ),
+    Morpheme(lemma="だ", inflection="だ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="だっ", inflection="だっ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="ああ", inflection="ああ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="ず", inflection="ず", highest_inflection_learning_interval=10),
+    Morpheme(lemma="た", inflection="た", highest_inflection_learning_interval=10),
+    Morpheme(lemma="相", inflection="相", highest_inflection_learning_interval=10),
+]
+
+##############################################################################################
+#                                    CASE: Morph and ruby interaction
+##############################################################################################
+# This third example sets all the cases where morphs and rubies coexist.
+# |-mmm---| |--mmm--| |---mmm-| |-mmm---| |-mmmmm-| |-mmmmm-| |--mmm--| |--m-m--|
+# |----rrr| |--rrr--| |--rrr--| |--rrr--| |--rrr--| |--r-r--| |-rrrrr-| |-rrrrr-|
+# Collection choice is arbitrary.
+# Database choice is arbitrary.
+##############################################################################################
```

Comment:
This shows all of the possible interactions between morphs and rubies. each of these has a test below.